tests/test_recursive_type.py

```diff
@@ -0,0 +1,159 @@
+import pytest
+from pydantic_core import ValidationError
+from typing_extensions import Annotated, TypeAlias
+
+from pydantic import BaseModel, RecursiveType
+
+Json: TypeAlias = list['Json'] | dict[str, 'Json'] | str | int | float | bool | None
+
+
+def test_recursive_type():
+    class Model(BaseModel):
+        x: Annotated[Json, RecursiveType('Json')]
```

Comment:
The reason the `RecursiveType('Json')` thing is required is because you have to specify the name _somewhere_ that pydantic can introspect.